### PR TITLE
add keytool_parse to assemblyline-v4-service instead of assemblyline-…

### PR DESCRIPTION
**Pull request by cccs-dawson**
Move keytool_parse--see [assemblyline-base PR](https://github.com/CybercentreCanada/assemblyline-base/pull/169)

### CODE REVIEW

**Comment by cccs-rs** [comment]
Might be a good idea to change the Dockerfile to include the installation for keytool since the base image doesn't currently contain it.

From APKaye Dockerfile:
```
# The following line fix an issue with openjdk installation
RUN mkdir -p /usr/share/man/man1

# Get required apt packages
RUN apt-get install -y openjdk-8-jre-headless java-common
```

Then we could remove those similar lines from the related services' Dockerfiles (namely APKaye and Espresso)

File: assemblyline_v4_service/common/keytool_parse.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    """
+    This function parses the output of 'keytool -printcert' and creates a list of Certificate objects.
+    The input to this function is the output of keytool_printcert
+
+    :param printcert: the string output of 'keytool -printcert'
+    :return: a list of the parsed out certificates. If only one certificate is present and not a chain, then the list will have one element.
+    """
+    certs = []
+
+    for cert_str in  re.split('Certificate\[\d+\]:', printcert): # split printcert output incase of certificate chain
```
Comment (cccs-kevin) [inline assemblyline_v4_service/common/keytool_parse.py:41]: `re.split(...) -> split(...)` once the import is changed

Comment (cccs-kevin) [inline assemblyline_v4_service/common/keytool_parse.py:41]: `for cert_str in  re.split -> for cert_str in re.split` (remove extra whitespace between `in` and `re.split`)

File: assemblyline_v4_service/common/keytool_parse.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        if cert_str == '':
+            continue
+        cert = Certificate()
+        cert.raw = cert_str.strip()
+        for line in cert_str.splitlines():
+            if "Owner:" in line:
```
Comment (cccs-kevin) [inline assemblyline_v4_service/common/keytool_parse.py:47]: is it possible for a line to have `Owner:`, `Issuer:` and `Valid from:` in it? If not then these consecutive `if` statements should be converted to `if, elif, elif`

**Review by cccs-kevin — CHANGES_REQUESTED** [review]

**Comment by cccs-kevin** [comment]
> 
> 
> Might be a good idea to change the Dockerfile to include the installation for keytool since the base image doesn't currently contain it.
> 
> From APKaye Dockerfile:
> 
> ```
> # The following line fix an issue with openjdk installation
> RUN mkdir -p /usr/share/man/man1
> 
> # Get required apt packages
> RUN apt-get install -y openjdk-8-jre-headless java-common
> ```
> 
> Then we could remove those similar lines from the related services' Dockerfiles (namely APKaye and Espresso)

Doesn't that make every service more heavyweight though?

**Comment by cccs-rs** [comment]
> > Might be a good idea to change the Dockerfile to include the installation for keytool since the base image doesn't currently contain it.
> > From APKaye Dockerfile:
> > ```
> > # The following line fix an issue with openjdk installation
> > RUN mkdir -p /usr/share/man/man1
> > 
> > # Get required apt packages
> > RUN apt-get install -y openjdk-8-jre-headless java-common
> > ```
> > 
> > 
> > Then we could remove those similar lines from the related services' Dockerfiles (namely APKaye and Espresso)
> 
> Doesn't that make every service unnecessarily more heavyweight though?

This is true, I guess it has to be noted somewhere that if a service wants to call keytool_parse.keytool_printcert() they need to have the library installed.

**Comment by cccs-kevin** [comment]
> 
> 
> > > Might be a good idea to change the Dockerfile to include the installation for keytool since the base image doesn't currently contain it.
> > > From APKaye Dockerfile:
> > > ```
> > > # The following line fix an issue with openjdk installation
> > > RUN mkdir -p /usr/share/man/man1
> > > 
> > > # Get required apt packages
> > > RUN apt-get install -y openjdk-8-jre-headless java-common
> > > ```
> > > 
> > > 
> > > Then we could remove those similar lines from the related services' Dockerfiles (namely APKaye and Espresso)
> > 
> > 
> > Doesn't that make every service unnecessarily more heavyweight though?
> 
> This is true, I guess it has to be noted somewhere that if a service wants to call something in keytool_parse they need to have the library installed.

We can just throw the sample Docker command in a docstring in `keytool_parse` and hopefully service developers read through the tools that they are trying to use haha

**Comment by cccs-kevin** [comment]
unit testing script for tool incoming 🚀

**Comment by cccs-sgaron** [comment]
Well then if only one service use this why is this not just part of that service?

**Comment by cccs-dawson** [comment]
this module is here and not in the espresso service because the apkaye service also uses keytool to parse the apks certificate, so it to could use this module

**Comment by cccs-rs** [comment]
Since keytool_parse is only used by a small subset of current services, we're electing to not make this part of the base service image (otherwise other services will be carrying unneeded bloat). 

Implementation will have to move to the specific service(s) that will use it, namely at the moment APKaye and Espresso.

**Comment by cccs-rs** [comment]
Seems this PR makes no changes to the Dockerfile which was the initial concern. 

Since there's only been a module addition, I see no reason for this to not be a common module for services (although there should be a note somewhere that the service needs to have keytool installed to leverage this module)

**Review by cccs-rs — APPROVED** [review]

**Review by cccs-kevin — APPROVED** [review]

**Pull request merged**